File: 2023/runall.py
#! /usr/bin/python3

import os
import subprocess
import time

# python c++ go

# Python is slowest, then unoptimized C++, then go and optimized C++ are about the same.

def fetch(cmd):
    return subprocess.run( cmd, check=True, stdout=subprocess.PIPE).stdout.decode()

def prep(file, ext):
    if ext == '.cpp':
        subprocess.run(['g++', '--std=c++17', '-O3', '-o', file, file+ext, '-llapack'])
    elif ext == '.go':
        subprocess.run(['go', 'build', file+ext] )

def run(file, ext):
    if ext == '.py':
        return fetch(['python', fn+ext])
    else:
        return fetch(['./'+fn])

print('Python'.ljust(30), 'C++'.ljust(30), 'Go')
print('-'*92)

for day in range(1,26):
    fn = f'day{day:02d}'
    gather = []
    times = []
    for lang in '.py','.cpp','.go':
        if os.path.exists(fn+lang):
            print(lang,end=' \r')
            prep( fn, lang )
            before = time.time()
            s = run( fn, lang )
            times.append( time.time() - before )
            gather.append( s.splitlines() )

    print(fn)
    pad = ''
    for lns in zip(*gather):
        for i,ln in enumerate(lns):
            print(pad,end='')
            if len(ln) > 30:
                print(ln)
                pad = ' '*(31*(i+1))
            else:
                print(ln.ljust(31),end='')
                pad = ''
        print()
    for t in times:
        print( ('%10.3f'%t).ljust(31), end='')
    print()

File: 2023/runall.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"strings"
	"time"
)

func ljust(s string, w int) string {
	if len(s) >= w {
		return s
	}
	return s + strings.Repeat(" ", w-len(s))
}

func fetch(cmd string, args ...string) string {
	sub := exec.Command(cmd, args...)
	output, err := sub.Output()
	if err != nil {
		fmt.Println("Error", err, "calling", cmd)
		return "Error"
	}
	return string(output)
}

func prep(file string, ext string) {
	if ext == ".cpp" {
		os.Remove(file)
		sub := exec.Command("g++", "--std=c++17", "-O3", "-o", file, file+ext, "-llapack")
		sub.Stdout = os.Stdout
		sub.Stderr = os.Stderr
		sub.Run()
	} else if ext == ".go" {
		os.Remove(file)
		sub := exec.Command("go", "build", file+ext)
		sub.Stdout = os.Stdout
		sub.Stderr = os.Stderr
		sub.Run()
	}
}

func run(file string, ext string) string {
	s := ""
	if ext == ".py" {
		s = fetch("python", file+ext)
	} else {
		s = fetch("./" + file)
	}
	return s
}

const FIELD int = 40

func main() {
	stats_csv, _ := os.Create("times.csv")
	defer stats_csv.Close()
	fmt.Println(ljust("Python", FIELD), ljust("C++", FIELD), "Go")
	fmt.Println(strings.Repeat("-", 3*FIELD))

	sums := []float64{0,0,0}
	for day := 1; day <= 25; day++ {
		fn := fmt.Sprintf("day%02d", day)
		fmt.Println(fn)

		var gather [][]string
		var times []time.Duration
		for _, lang := range []string{".py", ".cpp", ".go"} {
			if _, err := os.Stat(fn + lang); err == nil {
				prep(fn, lang)
				before := time.Now()
				s := run(fn, lang)
				times = append(times, time.Since(before))
				gather = append(gather, strings.Split(s, "\n"))
			}
		}

		pad := ""
		for i := 0; i < len(gather[0]); i++ {
			pad = ""
			if len(gather[0][i]) < 1 {
				break
			}
			for j, ln := range gather {
				fmt.Print(pad)
				if i >= len(ln) {
					fmt.Print(strings.Repeat(" ", FIELD))
				} else if len(ln[i]) > FIELD {
					fmt.Println(ln[i])
					pad = strings.Repeat(" ", (j+1)*FIELD)
				} else {
					fmt.Print(ljust(ln[i], FIELD))
					pad = ""
				}
			}
			if len(pad) == 0 {
				fmt.Println()
			}
		}
		if strings.Join(gather[0],"") != strings.Join(gather[1],"") {
			fmt.Println("#### Python and C++ results do not match ####")
		}
		if strings.Join(gather[0],"") != strings.Join(gather[2],"") {
			fmt.Println("#### Python and Go results do not match ####")
		}
		for i, d := range times {
			sums[i] += d.Seconds()
			fmt.Printf("%10.3f", d.Seconds())
			fmt.Print(strings.Repeat(" ", FIELD-10))
			if i > 0 {
				stats_csv.Write([]byte(","))
			}
			fmt.Fprint(stats_csv, d.Seconds())
		}
		fmt.Println()
		stats_csv.Write([]byte("\n"))
	}
    fmt.Println("\nTotals:")
	for _, d := range sums {
		fmt.Printf("%10.3f s", d)
		fmt.Print(strings.Repeat(" ", FIELD-12))
	}
	fmt.Println()
}
